src/basecamp.cpp

```diff
@@ -23,24 +74,159 @@ std::string basecamp::board_name() const
-    return data.str();
+bool basecamp::any_has_level( const std::string &type, int min_level ) const
+{
+    for( auto &expansion : expansions ) {
+        if( expansion.second.type == type && expansion.second.cur_level >= min_level ) {
+            return true;
+        }
+    }
+    return false;
 }
 
+bool basecamp::can_expand() const
+{
+    auto e = expansions.find( base_dir );
+    if( e == expansions.end() ) {
+        return false;
+    }
+    return static_cast<int>( directions.size() ) < ( e->second.cur_level / 2 - 1 );
+}
+
+const std::string basecamp::next_upgrade( const std::string &dir ) const
+{
+    auto e = expansions.find( dir );
+    if( e != expansions.end() ) {
+        int cur_level = e->second.cur_level;
+        if( cur_level >= 0 && cur_level < max_upgrade_by_type( e->second.type ) ) {
+            return faction_encode_full( e->second, 1 );
+        }
+    }
+    return "null";
+}
+
+// recipes and craft support functions
+std::map<std::string, std::string> basecamp::recipe_deck( const std::string &dir ) const
+{
+    if( dir == "ALL" || dir == "COOK" || dir == "BASE" || dir == "FARM" || dir == "SMITH" ) {
+        return recipe_group::get_recipes( dir );
+    }
+    std::map<std::string, std::string> cooking_recipes;
+    auto e = expansions.find( dir );
+    if( e == expansions.end() ) {
+        return cooking_recipes;
+    }
+    const std::string building = faction_encode( e->second.type );
+    int building_max = max_upgrade_by_type( e->second.type );
+    for( int building_cur = e->second.cur_level; building_cur <= building_max; building_cur++ ) {
+        const std::string building_level = building + to_string( building_cur );
+        if( !oter_str_id( building_level ) ) {
+            continue;
+        }
+        std::map<std::string, std::string> test_s = recipe_group::get_recipes( building_level );
+        cooking_recipes.insert( test_s.begin(), test_s.end() );
+    }
+    return cooking_recipes;
+}
+
+const std::string basecamp::get_gatherlist() const
+{
+    auto e = expansions.find( base_dir );
+    if( e != expansions.end() ) {
+        const std::string gatherlist = "gathering_" + faction_encode_full( e->second );
+        if( item_group::group_is_defined( gatherlist ) ) {
+            return gatherlist;
+        }
+    }
+    return "forest";
+
+}
+
+// display names
+std::string basecamp::expansion_tab( const std::string &dir ) const
+{
+    if( dir == base_dir ) {
+        return _( "Base Missions" );
+    }
+    auto e = expansions.find( dir );
+    if( e != expansions.end() ) {
+        if( e->second.type == "garage" ) {
+            return _( "Garage Expansion" );
+        }
+        if( e->second.type == "kitchen" ) {
+            return _( "Kitchen Expansion" );
+        }
+        if( e->second.type == "blacksmith" ) {
+            return _( "Blacksmith Expansion" );
+        }
+        if( e->second.type == "farm" ) {
+            return _( "Farm Expansion" );
+        }
```

Comment:
Small note that doesn't directly apply to this PR, but the camp manager dialogue still refers to building a "lumber mill" as the first recommended expansion. If these are the exhaustive list currently, I can go edit that to reflect it in a separate PR.

Reply:
Those are the current options.  There is no lumber mill though there should be.  It's on my to-do list as the first new expansion after I've gotten this stuff into a more flexible structure, but feel free to add it in the meantime or rewrite the dialogue as appropriate.

Reply:
I think the dialogue should change anyway, and that's something I know how to do without breaking the internet, so I'll get on that. As for adding it, I'm gettin' there... I understand how mapping works so I could at least draw up the maps for the construction stages.

Reply:
There, you got me editing cpp files now. I hope you're happy. https://github.com/CleverRaven/Cataclysm-DDA/pull/26887